src/liballoc/boxed.rs

```diff
@@ -63,6 +63,28 @@
 //! T` obtained from `Box::<T>::into_raw` may be deallocated using the
 //! [`Global`] allocator with `Layout::for_value(&*value)`.
 //!
+//! `Box<T>` has the same representation as `*mut T`. In particular, when
+//! `T: Sized`, this means that `Box<T>` has the same representation as
+//! a C pointer, making the following code valid in FFI:
+//!
+//! ```c
+//! /* C header */
+//! struct Foo* foo(); /* Returns ownership */
+//! void bar(struct Foo*); /* `bar` takes ownership */
+//! ```
+//!
+//! ```
+//! #[repr(C)]
+//! pub struct Foo;
+//!
+//! #[no_mangle]
+//! pub extern "C" fn foo() -> Box<Foo> {
+//!     Box::new(Foo)
+//! }
+//!
+//! #[no_mangle]
+//! pub extern "C" fn bar(_: Box<Foo>) {}
```

Comment:
I should probably change the parameter type of `bar` to `Option<Box<Foo>>` so that the Rust code can handle C code calling this function with a null pointer, which would lead to UB. I will update the PR tonight.

Reply:
I think that was fine-ish. This FFI already assumes lots of things about the pointer (that it's properly aligned, that it was allocated by Rust allocator and not something else, that the data behind pointer is also safe and doesn't contain invalid values etc. etc.), and nullability is not special enough IMO to single out and handle differently.